<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>singlecellmultiomics.molecule package &#8212; SingleCellMultiOmics 0.0.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="singlecellmultiomics-molecule-package">
<h1>singlecellmultiomics.molecule package<a class="headerlink" href="#singlecellmultiomics-molecule-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-singlecellmultiomics.molecule.molecule">
<span id="singlecellmultiomics-molecule-molecule-module"></span><h2>singlecellmultiomics.molecule.molecule module<a class="headerlink" href="#module-singlecellmultiomics.molecule.molecule" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="singlecellmultiomics.molecule.molecule.Molecule">
<em class="property">class </em><code class="descclassname">singlecellmultiomics.molecule.molecule.</code><code class="descname">Molecule</code><span class="sig-paren">(</span><em>fragments=None</em>, <em>cache_size=10000</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Molecule class, contains one or more associated fragments</p>
<dl class="attribute">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.fragments">
<code class="descname">fragments</code><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.fragments" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list</em> &#8211; associated fragments</p>
</dd></dl>

<dl class="attribute">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.spanStart">
<code class="descname">spanStart</code><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.spanStart" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> &#8211; starting coordinate of molecule, None if not available</p>
</dd></dl>

<dl class="attribute">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.spanEnd">
<code class="descname">spanEnd</code><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.spanEnd" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> &#8211; ending coordinate of molecule, None if not available</p>
</dd></dl>

<dl class="attribute">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.chromosome">
<code class="descname">chromosome</code><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> &#8211; mapping chromosome of molecule, None if not available</p>
</dd></dl>

<dl class="attribute">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.cache_size">
<code class="descname">cache_size</code><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.cache_size" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> &#8211; radius of molecule assignment cache</p>
</dd></dl>

<dl class="attribute">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.strand">
<code class="descname">strand</code><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.strand" title="Permalink to this definition">¶</a></dt>
<dd><p><em>bool</em> &#8211; mapping strand.
True when strand is REVERSE
False when strand is FORWARD
None when strand is not determined</p>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.__getitem__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtain a fragment belonging to this molecule.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> (<em>int</em>) &#8211; index of the fragment [0 ,1 , 2 ..]</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">fragment (singlecellmultiomics.fragment.Fragment)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.__iter__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.__iter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over all associated fragments</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Yields:</th><td class="field-body">singlecellmultiomics.fragment.Fragment</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.add_fragment">
<code class="descname">add_fragment</code><span class="sig-paren">(</span><em>fragment</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.add_fragment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.add_fragment" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.can_be_yielded">
<code class="descname">can_be_yielded</code><span class="sig-paren">(</span><em>chromosome</em>, <em>position</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.can_be_yielded"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.can_be_yielded" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the molecule is far enough away from the supplied location to be ejected from a buffer.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chromosome</strong> (<em>str</em>) &#8211; chromosome / contig of location to test</li>
<li><strong>position</strong> (<em>int</em>) &#8211; genomic location of location to test</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">can_be_yielded (bool)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.check_variants">
<code class="descname">check_variants</code><span class="sig-paren">(</span><em>variants</em>, <em>exclude_other_calls=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.check_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.check_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Verify variants in molecule</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>variants</strong> (<em>pysam.VariantFile</em>) &#8211; Variant file handle to extract variants from</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">{ (chrom,pos) : ( call (str) ): observations  (int) }</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict (collections.defaultdict( collections.Counter ))</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_TAPS_methylation_calls">
<code class="descname">get_TAPS_methylation_calls</code><span class="sig-paren">(</span><em>capture_context=None</em>, <em>reference=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_TAPS_methylation_calls"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_TAPS_methylation_calls" title="Permalink to this definition">¶</a></dt>
<dd><p>Get dictionary of TAPS methylated bases
:param capture_context:
:type capture_context: bool
:param reference:
:type reference: pysam.FastaFile</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">{ tuple(chrom,pos):  converted (bool) }</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">methylation_status (dict)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_base_observation_dict">
<code class="descname">get_base_observation_dict</code><span class="sig-paren">(</span><em>return_refbases=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_base_observation_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_base_observation_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtain observed bases at reference aligned locations</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>return_refbases</strong> (<em>bool</em>) &#8211; return both observed bases and reference bases</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">base (string) : obs (int) }
and
{ genome_location (tuple) : base (string) if return_refbases is True }</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">{ genome_location (tuple)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_consensus">
<code class="descname">get_consensus</code><span class="sig-paren">(</span><em>base_obs=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_consensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_consensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Get dictionary containing consensus calls in respect to reference</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>base_obs</strong> (<em>collections.defaultdict</em><em>(</em><em>collections.Counter</em><em>)</em>) &#8211; { genome_location (tuple) : base (string) : obs (int) }</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">{location : base}</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">consensus (dict)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_cut_site">
<code class="descname">get_cut_site</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_cut_site"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_cut_site" title="Permalink to this definition">¶</a></dt>
<dd><p>For restriction based protocol data, obtain genomic location of cut site</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None if site is not available<p>chromosome (str)
position (int)
strand (bool)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_mean_rt_fragment_size">
<code class="descname">get_mean_rt_fragment_size</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_mean_rt_fragment_size"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_mean_rt_fragment_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtain the mean RT reaction fragment size</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">mean_rt_size (float)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_methylated_count">
<code class="descname">get_methylated_count</code><span class="sig-paren">(</span><em>context=3</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_methylated_count"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_methylated_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the total amount of methylated bases
:param context: 3 or 4 base context
:type context: int</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">sum of methylated bases in contexts</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">r (collections.Counter)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_methylation_dict">
<code class="descname">get_methylation_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_methylation_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_methylation_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_rt_reaction_fragment_sizes">
<code class="descname">get_rt_reaction_fragment_sizes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_rt_reaction_fragment_sizes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_rt_reaction_fragment_sizes" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtain all RT reaction fragment sizes
:returns: rt_sizes (list of ints)</p>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_sample">
<code class="descname">get_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtain sample</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Sample associated with the molecule. Usually extracted from SM tag.
Calls fragment.get_sample() to obtain the sample</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">sample (str)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_strand">
<code class="descname">get_strand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_strand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtain mapping strand of molecule
:returns:</p>
<blockquote>
<div><dl class="docutils">
<dt>True,False,None</dt>
<dd>True when strand is REVERSE
False when strand is FORWARD
None when strand is not determined</dd>
</dl>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#singlecellmultiomics.molecule.molecule.Molecule.strand" title="singlecellmultiomics.molecule.molecule.Molecule.strand">strand</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.get_umi">
<code class="descname">get_umi</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.get_umi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.get_umi" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.is_multimapped">
<code class="descname">is_multimapped</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.is_multimapped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.is_multimapped" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="singlecellmultiomics.molecule.molecule.Molecule.iter_reads">
<code class="descname">iter_reads</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#Molecule.iter_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.Molecule.iter_reads" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over all associated reads
:returns: generator (pysam.AlignedSegment)</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="singlecellmultiomics.molecule.molecule.MoleculeIterator">
<code class="descclassname">singlecellmultiomics.molecule.molecule.</code><code class="descname">MoleculeIterator</code><span class="sig-paren">(</span><em>alignments</em>, <em>moleculeClass=&lt;class 'singlecellmultiomics.molecule.molecule.Molecule'&gt;</em>, <em>fragmentClass=&lt;class 'singlecellmultiomics.fragment.fragment.Fragment'&gt;</em>, <em>check_eject_every=1000</em>, <em>molecule_class_args={}</em>, <em>fragment_class_args={}</em>, <em>**pysamArgs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#MoleculeIterator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.MoleculeIterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over molecules in pysam.AlignmentFile</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>alignments</strong> (<em>pysam.AlignmentFile</em>) &#8211; Alignments to extract molecules form</li>
<li><strong>moleculeClass</strong> (<em>pysam.FastaFile</em>) &#8211; Class to use for molecules.</li>
<li><strong>fragmentClass</strong> (<em>pysam.FastaFile</em>) &#8211; Class to use for fragments.</li>
<li><strong>check_eject_every</strong> (<em>int</em>) &#8211; Check for yielding every N reads.</li>
<li><strong>molecule_class_args</strong> (<em>dict</em>) &#8211; arguments to pass to moleculeClass.</li>
<li><strong>fragment_class_args</strong> (<em>dict</em>) &#8211; arguments to pass to fragmentClass.</li>
<li><strong>**kwargs</strong> &#8211; arguments to pass to the pysam.AlignmentFile.fetch function</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Yields:</th><td class="field-body"><p class="first last"><em>methylation_status (dict)</em> &#8211; { tuple(chrom,pos):  converted (bool) }</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="singlecellmultiomics.molecule.molecule.molecule_to_random_primer_dict">
<code class="descclassname">singlecellmultiomics.molecule.molecule.</code><code class="descname">molecule_to_random_primer_dict</code><span class="sig-paren">(</span><em>molecule</em>, <em>primer_length=6</em>, <em>primer_read=2</em>, <em>max_N_distance=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singlecellmultiomics/molecule/molecule.html#molecule_to_random_primer_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#singlecellmultiomics.molecule.molecule.molecule_to_random_primer_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-singlecellmultiomics.molecule">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-singlecellmultiomics.molecule" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">singlecellmultiomics.molecule package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-singlecellmultiomics.molecule.molecule">singlecellmultiomics.molecule.molecule module</a></li>
<li><a class="reference internal" href="#module-singlecellmultiomics.molecule">Module contents</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/source/singlecellmultiomics.molecule.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Buys de Barbanson, Jake Yeung.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="../_sources/source/singlecellmultiomics.molecule.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>