<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>singlecellmultiomics.universalBamTagger.scchic &#8212; SingleCellMultiOmics 0.0.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for singlecellmultiomics.universalBamTagger.scchic</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">singlecellmultiomics.universalBamTagger.digest</span> <span class="k">import</span> <span class="n">DigestFlagger</span>
<span class="kn">from</span> <span class="nn">singlecellmultiomics.tagtools</span> <span class="k">import</span> <span class="n">tagtools</span>
<span class="n">complement</span> <span class="o">=</span> <span class="nb">str</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="s1">&#39;ATGC&#39;</span><span class="p">,</span> <span class="s1">&#39;TACG&#39;</span><span class="p">)</span>
<div class="viewcode-block" id="ChicSeqFlagger"><a class="viewcode-back" href="../../../source/singlecellmultiomics.universalBamTagger.html#singlecellmultiomics.universalBamTagger.scchic.ChicSeqFlagger">[docs]</a><span class="k">class</span> <span class="nc">ChicSeqFlagger</span><span class="p">(</span> <span class="n">DigestFlagger</span> <span class="p">):</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">DigestFlagger</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span> <span class="p">)</span>

<div class="viewcode-block" id="ChicSeqFlagger.addSite"><a class="viewcode-back" href="../../../source/singlecellmultiomics.universalBamTagger.html#singlecellmultiomics.universalBamTagger.scchic.ChicSeqFlagger.addSite">[docs]</a>    <span class="k">def</span> <span class="nf">addSite</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reads</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">restrictionChrom</span><span class="p">,</span> <span class="n">restrictionPos</span><span class="p">,</span><span class="n">is_trimmed</span><span class="o">=</span><span class="kc">False</span> <span class="p">):</span>

        <span class="n">sample</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sampleTag</span><span class="p">)</span>
        <span class="n">umi</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">umiTag</span><span class="p">)</span>
        <span class="n">allele</span> <span class="o">=</span> <span class="kc">None</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alleleTag</span><span class="p">)</span> <span class="k">else</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alleleTag</span><span class="p">)</span>
        <span class="n">siteInfo</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span> <span class="p">[</span> <span class="n">x</span>  <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">strand</span><span class="p">,</span> <span class="n">allele</span><span class="p">,</span> <span class="n">umi</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">])</span>
        <span class="n">moleculeId</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">increaseAndRecordOversequencing</span><span class="p">(</span>  <span class="n">sample</span><span class="p">,</span>  <span class="n">restrictionChrom</span><span class="p">,</span> <span class="n">restrictionPos</span><span class="p">,</span> <span class="n">siteInfo</span><span class="o">=</span><span class="n">siteInfo</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">read</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setSiteOversequencing</span><span class="p">(</span> <span class="n">read</span><span class="p">,</span> <span class="n">moleculeId</span> <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setSiteCoordinate</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">restrictionPos</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">is_trimmed</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">setRecognizedSequence</span><span class="p">(</span><span class="n">read</span><span class="p">,</span><span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;lh&#39;</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">:</span> <span class="c1">#</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">setRecognizedSequence</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">complement</span><span class="p">))</span> <span class="c1"># the first two base, this should be A{A:80%, N:20%}, we take the complement because the reads captures the complement strand</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">setRecognizedSequence</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span> <span class="c1"># the last two bases</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setSource</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="s1">&#39;CHIC&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">allele</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">setAllele</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">allele</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">setStrand</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span> <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="mi">1</span> <span class="k">else</span> <span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;?&#39;</span><span class="p">))</span> <span class="c1">#Note: strand 1 is +</span></div>

<div class="viewcode-block" id="ChicSeqFlagger.digest"><a class="viewcode-back" href="../../../source/singlecellmultiomics.universalBamTagger.html#singlecellmultiomics.universalBamTagger.scchic.ChicSeqFlagger.digest">[docs]</a>    <span class="k">def</span> <span class="nf">digest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reads</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span><span class="o">!=</span><span class="mi">2</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span> <span class="c1"># Only made for mate pair</span>

        <span class="n">R1</span><span class="p">,</span><span class="n">R2</span> <span class="o">=</span> <span class="n">reads</span>
        <span class="k">if</span> <span class="n">R1</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">R2</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">addAlleleInfo</span><span class="p">([</span><span class="n">read</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span> <span class="k">if</span> <span class="n">read</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">])</span>

        <span class="sd">&quot;&quot;&quot; Valid configs:</span>
<span class="sd">        Observed read pair:</span>
<span class="sd">        T######## R1 ########## ^ ########## R2 ##########</span>
<span class="sd">        Real molecule:</span>
<span class="sd">        A###########################################################</span>

<span class="sd">        Real molecule: and adapter:</span>
<span class="sd">        ADAPTER: 5&#39; NNNNT 3&#39;</span>
<span class="sd">                        A{A:80%,N:20%}NNN [CHIC MOLECULE]</span>
<span class="sd">                                      ^ real cut site</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">is_trimmed</span> <span class="o">=</span> <span class="p">(</span><span class="n">R1</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s1">&#39;MX&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">R1</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;MX&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;scCHIC&#39;</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">R1</span><span class="o">.</span><span class="n">is_unmapped</span> <span class="ow">or</span> <span class="n">R2</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">:</span>
            <span class="k">return</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">tagtools</span><span class="o">.</span><span class="n">getPairGenomicLocations</span><span class="p">(</span><span class="n">R1</span><span class="p">,</span><span class="n">R2</span><span class="p">,</span> <span class="n">R1PrimerLength</span><span class="o">=</span><span class="mi">1</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">is_trimmed</span><span class="p">),</span> <span class="n">R2PrimerLength</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setFragmentSize</span><span class="p">(</span><span class="n">R1</span><span class="p">,</span> <span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setFragmentSize</span><span class="p">(</span><span class="n">R2</span><span class="p">,</span> <span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setFragmentTrust</span><span class="p">(</span><span class="n">R1</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setFragmentTrust</span><span class="p">(</span><span class="n">R2</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setFragmentSize</span><span class="p">(</span><span class="n">R1</span><span class="p">,</span> <span class="s1">&#39;unknown&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">setFragmentSize</span><span class="p">(</span><span class="n">R2</span><span class="p">,</span> <span class="s1">&#39;unknown&#39;</span><span class="p">)</span>


        <span class="c1">#if R1.seq[0]==&#39;T&#39;: # Site on the start of R1, R2 should map behind</span>
        <span class="k">if</span> <span class="n">is_trimmed</span><span class="p">:</span>
            <span class="c1"># The first base of the read has been taken off and the lh tag is already set, this can be copied to RZ</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">addSite</span><span class="p">(</span> <span class="p">[</span><span class="n">R1</span><span class="p">,</span><span class="n">R2</span><span class="p">],</span>
                <span class="n">strand</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">R1</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">),</span> <span class="c1"># We sequence the other strand (Starting with a T, this is an A in the molecule), the digestion thus happened on the other strand</span>
                <span class="c1"># On the next line we asume that the mnsase cut is one base after the ligated A, but it can be more bases upstream</span>
                <span class="n">restrictionChrom</span><span class="o">=</span><span class="n">R1</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span>
                <span class="n">restrictionPos</span><span class="o">=</span><span class="p">(</span><span class="n">R1</span><span class="o">.</span><span class="n">reference_end</span> <span class="k">if</span> <span class="n">R1</span><span class="o">.</span><span class="n">is_reverse</span> <span class="k">else</span> <span class="n">R1</span><span class="o">.</span><span class="n">reference_start</span><span class="p">),</span>
                <span class="n">is_trimmed</span><span class="o">=</span><span class="kc">True</span>

                  <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">addSite</span><span class="p">(</span> <span class="p">[</span><span class="n">R1</span><span class="p">,</span><span class="n">R2</span><span class="p">],</span>
                <span class="n">strand</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">R1</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">),</span> <span class="c1"># We sequence the other strand (Starting with a T, this is an A in the molecule), the digestion thus happened on the other strand</span>
                <span class="c1"># On the next line we asume that the mnsase cut is one base after the ligated A, but it can be more bases upstream</span>
                <span class="n">restrictionChrom</span><span class="o">=</span><span class="n">R1</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span>
                <span class="n">restrictionPos</span><span class="o">=</span><span class="p">(</span><span class="n">R1</span><span class="o">.</span><span class="n">reference_end</span><span class="o">-</span><span class="mi">1</span> <span class="k">if</span> <span class="n">R1</span><span class="o">.</span><span class="n">is_reverse</span> <span class="k">else</span> <span class="n">R1</span><span class="o">.</span><span class="n">reference_start</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>
                <span class="n">is_trimmed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span><span class="p">((</span> <span class="n">R1</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">R1</span><span class="o">.</span><span class="n">reference_start</span><span class="p">))</span></div></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Buys de Barbanson, Jake Yeung.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>