<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>singlecellmultiomics.molecule.molecule &#8212; SingleCellMultiOmics 0.0.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for singlecellmultiomics.molecule.molecule</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">singlecellmultiomics.utils.sequtils</span> <span class="k">import</span> <span class="n">hamming_distance</span>
<span class="kn">import</span> <span class="nn">pysamiterators.iterators</span>
<span class="kn">from</span> <span class="nn">singlecellmultiomics.fragment</span> <span class="k">import</span> <span class="n">Fragment</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<div class="viewcode-block" id="molecule_to_random_primer_dict"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.molecule_to_random_primer_dict">[docs]</a><span class="k">def</span> <span class="nf">molecule_to_random_primer_dict</span><span class="p">(</span><span class="n">molecule</span><span class="p">,</span> <span class="n">primer_length</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">primer_read</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">max_N_distance</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span> <span class="c1">#1: read1 2: read2</span>
    <span class="n">rp</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

    <span class="c1"># First add all reactions without a N in the sequence:</span>
    <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="n">molecule</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">fragment</span><span class="p">[</span><span class="n">primer_read</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">hstart</span><span class="p">,</span> <span class="n">hseq</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_random_primer_hash</span><span class="p">()</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="s1">&#39;N&#39;</span> <span class="ow">in</span> <span class="n">hseq</span><span class="p">:</span>
                <span class="n">rp</span><span class="p">[</span><span class="n">hstart</span><span class="p">,</span> <span class="n">hseq</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fragment</span><span class="p">)</span>

    <span class="c1"># Try to match reactions with N with reactions without a N</span>
    <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="n">molecule</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">fragment</span><span class="p">[</span><span class="n">primer_read</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">hstart</span><span class="p">,</span> <span class="n">hseq</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_random_primer_hash</span><span class="p">()</span>
            <span class="k">if</span> <span class="s1">&#39;N&#39;</span> <span class="ow">in</span> <span class="n">hseq</span><span class="p">:</span>
                <span class="c1"># find nearest</span>
                <span class="k">for</span> <span class="n">other_start</span><span class="p">,</span> <span class="n">other_seq</span> <span class="ow">in</span> <span class="n">rp</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">other_start</span><span class="o">!=</span><span class="n">hstart</span><span class="p">:</span>
                        <span class="k">continue</span>

                    <span class="k">if</span> <span class="s1">&#39;N&#39;</span> <span class="ow">in</span> <span class="n">other_seq</span><span class="p">:</span>
                        <span class="k">continue</span>

                    <span class="k">if</span> <span class="n">hamming_distance</span><span class="p">(</span><span class="n">hseq</span><span class="p">,</span><span class="n">other_seq</span><span class="p">)</span><span class="o">&lt;=</span><span class="n">max_N_distance</span><span class="p">:</span>
                        <span class="n">rp</span><span class="p">[</span><span class="n">other_start</span><span class="p">,</span> <span class="n">other_seq</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fragment</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">rp</span></div>

<div class="viewcode-block" id="Molecule"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule">[docs]</a><span class="k">class</span> <span class="nc">Molecule</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Molecule class, contains one or more associated fragments</span>

<span class="sd">    Attributes:</span>
<span class="sd">        fragments (list): associated fragments</span>

<span class="sd">        spanStart (int): starting coordinate of molecule, None if not available</span>

<span class="sd">        spanEnd (int): ending coordinate of molecule, None if not available</span>

<span class="sd">        chromosome (str): mapping chromosome of molecule, None if not available</span>

<span class="sd">        cache_size (int): radius of molecule assignment cache</span>

<span class="sd">        strand (bool): mapping strand.</span>
<span class="sd">            True when strand is REVERSE</span>
<span class="sd">            False when strand is FORWARD</span>
<span class="sd">            None when strand is not determined</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fragments</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cache_size</span><span class="o">=</span><span class="mi">10</span><span class="n">_000</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Initialise Molecule</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        fragments :  list(singlecellmultiomics.fragment.Fragment)</span>
<span class="sd">            Fragment to assign to Molecule. More fragments can be added later</span>

<span class="sd">        cache_size (int): radius of molecule assignment cache</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span>  <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spanStart</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spanEnd</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cache_size</span> <span class="o">=</span> <span class="n">cache_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">fragments</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">fragments</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">list</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">frag</span> <span class="ow">in</span> <span class="n">fragments</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">add_fragment</span><span class="p">(</span><span class="n">frag</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">add_fragment</span><span class="p">(</span><span class="n">fragments</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">)</span>



<div class="viewcode-block" id="Molecule.get_strand"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_strand">[docs]</a>    <span class="k">def</span> <span class="nf">get_strand</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Obtain mapping strand of molecule</span>
<span class="sd">        Returns:</span>
<span class="sd">            strand : True,False,None</span>
<span class="sd">                True when strand is REVERSE</span>
<span class="sd">                False when strand is FORWARD</span>
<span class="sd">                None when strand is not determined</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span></div>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">frag_repr</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">fragment</span><span class="p">)</span> <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">f</span><span class="s2">&quot;&quot;&quot;Molecule</span>
<span class="s2">        with {len(self.fragments)} assinged fragments</span>
<span class="s2">        &quot;&quot;&quot;</span> <span class="o">+</span> <span class="n">frag_repr</span>

<div class="viewcode-block" id="Molecule.get_umi"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_umi">[docs]</a>    <span class="k">def</span> <span class="nf">get_umi</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">umi_abundance</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">:</span>
            <span class="n">umi_abundance</span><span class="p">[</span><span class="n">fragment</span><span class="o">.</span><span class="n">get_umi</span><span class="p">()]</span><span class="o">+=</span><span class="mi">1</span>
        <span class="k">return</span> <span class="n">umi_abundance</span><span class="o">.</span><span class="n">most_common</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span></div>

<div class="viewcode-block" id="Molecule.get_sample"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_sample">[docs]</a>    <span class="k">def</span> <span class="nf">get_sample</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Obtain sample</span>

<span class="sd">        Returns:</span>
<span class="sd">            sample (str):</span>
<span class="sd">                Sample associated with the molecule. Usually extracted from SM tag.</span>
<span class="sd">                Calls fragment.get_sample() to obtain the sample</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_sample</span><span class="p">()</span></div>


<div class="viewcode-block" id="Molecule.get_cut_site"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_cut_site">[docs]</a>    <span class="k">def</span> <span class="nf">get_cut_site</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;For restriction based protocol data, obtain genomic location of cut site</span>

<span class="sd">        Returns:</span>
<span class="sd">            None if site is not available</span>

<span class="sd">            chromosome (str)</span>
<span class="sd">            position (int)</span>
<span class="sd">            strand (bool)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">:</span>
            <span class="n">site</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_site_location</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">site</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span> <span class="p">(</span><span class="o">*</span><span class="n">site</span><span class="p">,</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_strand</span><span class="p">()))</span>
        <span class="k">return</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="Molecule.is_multimapped"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.is_multimapped">[docs]</a>    <span class="k">def</span> <span class="nf">is_multimapped</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">fragment</span><span class="o">.</span><span class="n">is_multimapped</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="kc">True</span></div>

    <span class="k">def</span> <span class="nf">_add_fragment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fragment</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fragment</span><span class="p">)</span>
        <span class="c1"># Update span:</span>
        <span class="n">add_span</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_span</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spanStart</span> <span class="o">=</span> <span class="n">add_span</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">spanStart</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="nb">min</span><span class="p">(</span><span class="n">add_span</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">spanStart</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spanEnd</span> <span class="o">=</span> <span class="n">add_span</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">spanEnd</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="nb">max</span><span class="p">(</span><span class="n">add_span</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">spanEnd</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span> <span class="o">=</span> <span class="n">add_span</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">fragment</span><span class="o">.</span><span class="n">strand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">strand</span>

<div class="viewcode-block" id="Molecule.add_fragment"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.add_fragment">[docs]</a>    <span class="k">def</span> <span class="nf">add_fragment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fragment</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_fragment</span><span class="p">(</span><span class="n">fragment</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">True</span>

        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">f</span> <span class="o">==</span> <span class="n">fragment</span><span class="p">:</span>
                <span class="c1"># it matches this molecule:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_add_fragment</span><span class="p">(</span><span class="n">fragment</span><span class="p">)</span>
                <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span></div>

<div class="viewcode-block" id="Molecule.can_be_yielded"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.can_be_yielded">[docs]</a>    <span class="k">def</span> <span class="nf">can_be_yielded</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Check if the molecule is far enough away from the supplied location to be ejected from a buffer.</span>

<span class="sd">        Args:</span>
<span class="sd">            chromosome (str) : chromosome / contig of location to test</span>
<span class="sd">            position (int) : genomic location of location to test</span>

<span class="sd">        Returns:</span>
<span class="sd">            can_be_yielded (bool)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">chromosome</span><span class="o">!=</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spanStart</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">cache_size</span><span class="o">*</span><span class="mf">0.5</span><span class="p">)</span> <span class="ow">or</span> <span class="n">position</span> <span class="o">&gt;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spanEnd</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">cache_size</span><span class="o">*</span><span class="mf">0.5</span><span class="p">)</span></div>

<div class="viewcode-block" id="Molecule.get_rt_reaction_fragment_sizes"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_rt_reaction_fragment_sizes">[docs]</a>    <span class="k">def</span> <span class="nf">get_rt_reaction_fragment_sizes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Obtain all RT reaction fragment sizes</span>
<span class="sd">        Returns:</span>
<span class="sd">            rt_sizes (list of ints)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">rt_reactions</span> <span class="o">=</span> <span class="n">molecule_to_random_primer_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">amount_of_rt_reactions</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rt_reactions</span><span class="p">)</span>

        <span class="c1">#this obtains the maximum fragment size:</span>
        <span class="n">frag_chrom</span><span class="p">,</span> <span class="n">frag_start</span><span class="p">,</span> <span class="n">frag_end</span> <span class="o">=</span> <span class="n">pysamiterators</span><span class="o">.</span><span class="n">iterators</span><span class="o">.</span><span class="n">getListSpanningCoordinates</span><span class="p">([</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">])</span>

        <span class="c1">#Obtain the fragment sizes of all RT reactions:</span>
        <span class="n">rt_sizes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">rt_end</span><span class="p">,</span><span class="n">hexamer</span><span class="p">),</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="n">rt_reactions</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">rt_chrom</span><span class="p">,</span> <span class="n">rt_start</span><span class="p">,</span> <span class="n">rt_end</span> <span class="o">=</span> <span class="n">pysamiterators</span><span class="o">.</span><span class="n">iterators</span><span class="o">.</span><span class="n">getListSpanningCoordinates</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">fragment</span><span class="p">))</span>
            <span class="n">rt_sizes</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">rt_end</span><span class="o">-</span><span class="n">rt_start</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">rt_sizes</span></div>

<div class="viewcode-block" id="Molecule.get_mean_rt_fragment_size"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_mean_rt_fragment_size">[docs]</a>    <span class="k">def</span> <span class="nf">get_mean_rt_fragment_size</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Obtain the mean RT reaction fragment size</span>

<span class="sd">        Returns:</span>
<span class="sd">            mean_rt_size (float)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_rt_reaction_fragment_sizes</span><span class="p">()</span>
        <span class="p">)</span></div>



<div class="viewcode-block" id="Molecule.__getitem__"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.__getitem__">[docs]</a>    <span class="k">def</span> <span class="nf">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Obtain a fragment belonging to this molecule.</span>

<span class="sd">        Args:</span>
<span class="sd">            index (int):</span>
<span class="sd">                index of the fragment [0 ,1 , 2 ..]</span>

<span class="sd">        Returns:</span>
<span class="sd">            fragment (singlecellmultiomics.fragment.Fragment)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">[</span><span class="n">key</span><span class="p">]</span></div>



<div class="viewcode-block" id="Molecule.get_base_observation_dict"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_base_observation_dict">[docs]</a>    <span class="k">def</span> <span class="nf">get_base_observation_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">return_refbases</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Obtain observed bases at reference aligned locations</span>

<span class="sd">        Args:</span>
<span class="sd">            return_refbases ( bool ):</span>
<span class="sd">                return both observed bases and reference bases</span>

<span class="sd">        Returns:</span>
<span class="sd">            { genome_location (tuple) : base (string) : obs (int) }</span>
<span class="sd">            and</span>
<span class="sd">            { genome_location (tuple) : base (string) if return_refbases is True }</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">base_obs</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">return_refbases</span><span class="p">:</span>
            <span class="n">ref_bases</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="n">R1</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_R1</span><span class="p">()</span>
            <span class="n">R2</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_R2</span><span class="p">()</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">pysamiterators</span><span class="o">.</span><span class="n">iterators</span><span class="o">.</span><span class="n">getPairGenomicLocations</span><span class="p">(</span>
                <span class="n">R1</span><span class="o">=</span><span class="n">R1</span><span class="p">,</span>
                <span class="n">R2</span><span class="o">=</span><span class="n">R2</span><span class="p">,</span>
                <span class="n">allow_unsafe</span><span class="o">=</span><span class="p">(</span><span class="n">R1</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">))</span>

            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="p">[</span><span class="n">R1</span><span class="p">,</span><span class="n">R2</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">read</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">for</span> <span class="n">cycle</span><span class="p">,</span> <span class="n">query_pos</span><span class="p">,</span> <span class="n">ref_pos</span><span class="p">,</span> <span class="n">ref_base</span> <span class="ow">in</span> <span class="n">pysamiterators</span><span class="o">.</span><span class="n">iterators</span><span class="o">.</span><span class="n">ReadCycleIterator</span><span class="p">(</span><span class="n">read</span><span class="p">,</span><span class="n">with_seq</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">query_pos</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ref_pos</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ref_pos</span><span class="o">&lt;</span><span class="n">start</span> <span class="ow">or</span> <span class="n">ref_pos</span><span class="o">&gt;</span><span class="n">end</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">query_base</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">query_pos</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">query_base</span><span class="o">==</span><span class="s1">&#39;N&#39;</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">base_obs</span><span class="p">[(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span><span class="n">ref_pos</span><span class="p">)][</span><span class="n">query_base</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="k">if</span> <span class="n">return_refbases</span><span class="p">:</span>
                        <span class="n">ref_bases</span><span class="p">[(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span><span class="n">ref_pos</span><span class="p">)]</span><span class="o">=</span><span class="n">ref_base</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">return_refbases</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">base_obs</span><span class="p">,</span> <span class="n">ref_bases</span>
        <span class="k">return</span> <span class="n">base_obs</span></div>



<div class="viewcode-block" id="Molecule.get_consensus"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_consensus">[docs]</a>    <span class="k">def</span> <span class="nf">get_consensus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">base_obs</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get dictionary containing consensus calls in respect to reference</span>

<span class="sd">        Args:</span>
<span class="sd">            base_obs (collections.defaultdict(collections.Counter)) :</span>
<span class="sd">                { genome_location (tuple) : base (string) : obs (int) }</span>

<span class="sd">        Returns:</span>
<span class="sd">            consensus (dict)  :  {location : base}</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">consensus</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># postion -&gt; base , key is not set when not decided</span>
        <span class="k">if</span> <span class="n">base_obs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">base_obs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_base_observation_dict</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">location</span><span class="p">,</span> <span class="n">obs</span> <span class="ow">in</span> <span class="n">base_obs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">votes</span> <span class="o">=</span> <span class="n">obs</span><span class="o">.</span><span class="n">most_common</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">votes</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span> <span class="ow">or</span> <span class="n">votes</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">&lt;</span><span class="n">votes</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]:</span>
                <span class="n">consensus</span><span class="p">[</span><span class="n">location</span><span class="p">]</span> <span class="o">=</span> <span class="n">votes</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">consensus</span></div>

<div class="viewcode-block" id="Molecule.check_variants"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.check_variants">[docs]</a>    <span class="k">def</span> <span class="nf">check_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variants</span><span class="p">,</span> <span class="n">exclude_other_calls</span><span class="o">=</span><span class="kc">True</span> <span class="p">):</span> <span class="c1">#when enabled other calls (non ref non alt will be set None)</span>
        <span class="sd">&quot;&quot;&quot;Verify variants in molecule</span>

<span class="sd">        Args:</span>
<span class="sd">            variants (pysam.VariantFile) : Variant file handle to extract variants from</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict (collections.defaultdict( collections.Counter )) : { (chrom,pos) : ( call (str) ): observations  (int) }</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variant_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">spanStart</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">spanEnd</span> <span class="p">):</span>
            <span class="n">variant_dict</span><span class="p">[</span> <span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">pos</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span> <span class="o">=</span> <span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">alts</span><span class="p">)</span>


        <span class="n">variant_calls</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span> <span class="p">)</span>
        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>

            <span class="n">R1</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_R1</span><span class="p">()</span>
            <span class="n">R2</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_R2</span><span class="p">()</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">pysamiterators</span><span class="o">.</span><span class="n">iterators</span><span class="o">.</span><span class="n">getPairGenomicLocations</span><span class="p">(</span>
                <span class="n">R1</span><span class="o">=</span><span class="n">R1</span><span class="p">,</span>
                <span class="n">R2</span><span class="o">=</span><span class="n">R2</span><span class="p">,</span>
                <span class="n">allow_unsafe</span><span class="o">=</span><span class="p">(</span><span class="n">R1</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">end</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">fragment</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">read</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="k">for</span> <span class="n">cycle</span><span class="p">,</span> <span class="n">query_pos</span><span class="p">,</span> <span class="n">ref_pos</span> <span class="ow">in</span> <span class="n">pysamiterators</span><span class="o">.</span><span class="n">iterators</span><span class="o">.</span><span class="n">ReadCycleIterator</span><span class="p">(</span><span class="n">read</span><span class="p">):</span>

                    <span class="k">if</span> <span class="n">query_pos</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ref_pos</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ref_pos</span><span class="o">&lt;</span><span class="n">start</span> <span class="ow">or</span> <span class="n">ref_pos</span><span class="o">&gt;</span><span class="n">end</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">query_base</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">query_pos</span><span class="p">]</span>

                    <span class="n">k</span> <span class="o">=</span>  <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">ref_pos</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">variant_dict</span><span class="p">:</span>
                        <span class="n">call</span> <span class="o">=</span> <span class="kc">None</span>
                        <span class="n">ref</span><span class="p">,</span> <span class="n">alts</span> <span class="o">=</span> <span class="n">variant_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">query_base</span> <span class="o">==</span> <span class="n">ref</span><span class="p">:</span>
                            <span class="n">call</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;ref&#39;</span><span class="p">,</span><span class="n">query_base</span><span class="p">)</span>
                        <span class="k">elif</span> <span class="n">query_base</span> <span class="ow">in</span> <span class="n">alts</span><span class="p">:</span>
                            <span class="n">call</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;alt&#39;</span><span class="p">,</span><span class="n">query_base</span><span class="p">)</span>

                        <span class="k">if</span> <span class="ow">not</span> <span class="n">exclude_other_calls</span> <span class="ow">or</span> <span class="n">call</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                            <span class="n">variant_calls</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">call</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>

        <span class="k">return</span> <span class="n">variant_calls</span></div>


<div class="viewcode-block" id="Molecule.iter_reads"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.iter_reads">[docs]</a>    <span class="k">def</span> <span class="nf">iter_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterate over all associated reads</span>
<span class="sd">        Returns:</span>
<span class="sd">            generator (pysam.AlignedSegment)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">fragment</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">read</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">yield</span> <span class="n">read</span></div>

<div class="viewcode-block" id="Molecule.__iter__"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.__iter__">[docs]</a>    <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterate over all associated fragments</span>

<span class="sd">        Yields:</span>
<span class="sd">            singlecellmultiomics.fragment.Fragment</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">fragment</span></div>


<div class="viewcode-block" id="Molecule.get_methylated_count"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_methylated_count">[docs]</a>    <span class="k">def</span> <span class="nf">get_methylated_count</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">context</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get the total amount of methylated bases</span>
<span class="sd">        Args:</span>
<span class="sd">            context (int) : 3 or 4 base context</span>

<span class="sd">        Returns:</span>
<span class="sd">            r (collections.Counter) : sum of methylated bases in contexts</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">r</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">()</span></div>


<div class="viewcode-block" id="Molecule.get_TAPS_methylation_calls"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_TAPS_methylation_calls">[docs]</a>    <span class="k">def</span> <span class="nf">get_TAPS_methylation_calls</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">capture_context</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get dictionary of TAPS methylated bases</span>
<span class="sd">        Args:</span>
<span class="sd">            capture_context (bool)</span>

<span class="sd">            reference (pysam.FastaFile)</span>

<span class="sd">        Returns:</span>
<span class="sd">            methylation_status (dict): { tuple(chrom,pos):  converted (bool) }</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_multimapped</span><span class="p">():</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">strand</span> <span class="o">=</span>  <span class="bp">self</span><span class="o">.</span><span class="n">get_strand</span><span class="p">()</span> <span class="c1">#molecule.fragments[0][0].get_tag(&#39;RS&#39;)==&#39;-&#39;</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">base_obs</span><span class="p">,</span> <span class="n">ref_bases</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_base_observation_dict</span><span class="p">(</span><span class="n">return_refbases</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="c1"># We cannot determine a reliable consensus sequence</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">consensus</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_consensus</span><span class="p">(</span><span class="n">base_obs</span><span class="o">=</span><span class="n">base_obs</span><span class="p">)</span>
        <span class="n">methylation_status</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># location-&gt; call</span>
        <span class="k">for</span> <span class="n">location</span><span class="p">,</span> <span class="n">ref_base</span> <span class="ow">in</span> <span class="n">ref_bases</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">location</span> <span class="ow">in</span> <span class="n">consensus</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">query_base</span> <span class="o">=</span> <span class="n">consensus</span><span class="p">[</span><span class="n">location</span><span class="p">]</span>
            <span class="n">k</span> <span class="o">=</span> <span class="p">(</span><span class="o">*</span><span class="n">location</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ref_base</span><span class="o">==</span><span class="s1">&#39;C&#39;</span> <span class="ow">and</span> <span class="n">strand</span><span class="p">:</span> <span class="c1"># fwd</span>
                <span class="k">if</span> <span class="n">query_base</span><span class="o">==</span><span class="s1">&#39;T&#39;</span><span class="p">:</span>
                    <span class="n">methylation_status</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1">#  modified</span>
                <span class="k">elif</span> <span class="n">query_base</span><span class="o">==</span><span class="s1">&#39;C&#39;</span><span class="p">:</span>
                    <span class="n">methylation_status</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># not modified</span>

            <span class="k">elif</span> <span class="n">ref_base</span><span class="o">==</span><span class="s1">&#39;G&#39;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">strand</span><span class="p">:</span> <span class="c1"># rev</span>

                <span class="k">if</span> <span class="n">query_base</span><span class="o">==</span><span class="s1">&#39;A&#39;</span><span class="p">:</span>
                    <span class="n">methylation_status</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1">#  modified</span>
                <span class="k">elif</span> <span class="n">query_base</span><span class="o">==</span><span class="s1">&#39;G&#39;</span><span class="p">:</span>
                    <span class="n">methylation_status</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># not modified</span>

        <span class="k">return</span> <span class="n">methylation_status</span></div>


<div class="viewcode-block" id="Molecule.get_methylation_dict"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.Molecule.get_methylation_dict">[docs]</a>    <span class="k">def</span> <span class="nf">get_methylation_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">methylated_positions</span> <span class="o">=</span>  <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span> <span class="p">()</span> <span class="c1">#chrom-pos-&gt;count</span>
        <span class="n">methylated_state</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span><span class="c1">#chrom-pos-&gt;1, 0, -1</span>
        <span class="k">for</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">fragment</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">read</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s1">&#39;XM&#39;</span><span class="p">):</span>
                    <span class="k">continue</span>
                <span class="n">methylation_status_string</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;XM&#39;</span><span class="p">)</span>
                <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">qpos</span><span class="p">,</span> <span class="n">rpos</span><span class="p">,</span> <span class="n">ref_base</span> <span class="ow">in</span> <span class="n">read</span><span class="o">.</span><span class="n">get_aligned_pairs</span><span class="p">(</span><span class="n">with_seq</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">qpos</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="k">if</span> <span class="n">ref_base</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="k">if</span> <span class="n">rpos</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">methylation_status</span> <span class="o">=</span> <span class="n">methylation_status_string</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">methylation_status</span><span class="o">.</span><span class="n">isupper</span><span class="p">():</span>
                        <span class="n">methylated_positions</span><span class="p">[(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">rpos</span><span class="p">)]</span><span class="o">+=</span><span class="mi">1</span>
                        <span class="k">if</span> <span class="n">methylated_state</span><span class="o">.</span><span class="n">get</span><span class="p">(</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">rpos</span><span class="p">),</span><span class="mi">1</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span> <span class="p">:</span>
                            <span class="n">methylated_state</span><span class="p">[(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">rpos</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">methylated_state</span><span class="p">[(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">rpos</span><span class="p">)]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">methylation_status</span><span class="o">==</span><span class="s1">&#39;.&#39;</span><span class="p">:</span>
                            <span class="k">pass</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">methylated_state</span><span class="o">.</span><span class="n">get</span><span class="p">(</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">rpos</span><span class="p">),</span><span class="mi">0</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
                                <span class="n">methylated_state</span><span class="p">[(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">rpos</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">methylated_state</span><span class="p">[(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">rpos</span><span class="p">)]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                    <span class="n">i</span><span class="o">+=</span><span class="mi">1</span>
        <span class="k">return</span> <span class="n">methylated_positions</span><span class="p">,</span><span class="n">methylated_state</span></div></div>


<div class="viewcode-block" id="MoleculeIterator"><a class="viewcode-back" href="../../../source/singlecellmultiomics.molecule.html#singlecellmultiomics.molecule.molecule.MoleculeIterator">[docs]</a><span class="k">def</span> <span class="nf">MoleculeIterator</span><span class="p">(</span> <span class="n">alignments</span><span class="p">,</span> <span class="n">moleculeClass</span><span class="o">=</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">fragmentClass</span><span class="o">=</span><span class="n">Fragment</span><span class="p">,</span> <span class="n">check_eject_every</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">molecule_class_args</span><span class="o">=</span><span class="p">{},</span> <span class="n">fragment_class_args</span><span class="o">=</span><span class="p">{},</span> <span class="o">**</span><span class="n">pysamArgs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Iterate over molecules in pysam.AlignmentFile</span>

<span class="sd">    Args:</span>
<span class="sd">        alignments (pysam.AlignmentFile): Alignments to extract molecules form</span>

<span class="sd">        moleculeClass (pysam.FastaFile): Class to use for molecules.</span>

<span class="sd">        fragmentClass (pysam.FastaFile): Class to use for fragments.</span>

<span class="sd">        check_eject_every (int): Check for yielding every N reads.</span>

<span class="sd">        molecule_class_args (dict): arguments to pass to moleculeClass.</span>

<span class="sd">        fragment_class_args (dict): arguments to pass to fragmentClass.</span>

<span class="sd">        **kwargs: arguments to pass to the pysam.AlignmentFile.fetch function</span>

<span class="sd">    Yields:</span>
<span class="sd">        methylation_status (dict): { tuple(chrom,pos):  converted (bool) }</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">molecules</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">added_fragments</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">R1</span><span class="p">,</span><span class="n">R2</span> <span class="ow">in</span> <span class="n">pysamiterators</span><span class="o">.</span><span class="n">iterators</span><span class="o">.</span><span class="n">MatePairIterator</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span><span class="n">performProperPairCheck</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="o">**</span><span class="n">pysamArgs</span><span class="p">):</span>

        <span class="n">fragment</span> <span class="o">=</span> <span class="n">fragmentClass</span><span class="p">([</span><span class="n">R1</span><span class="p">,</span><span class="n">R2</span><span class="p">],</span> <span class="o">**</span><span class="n">fragment_class_args</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">fragment</span><span class="o">.</span><span class="n">is_valid</span><span class="p">()</span> <span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">added</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">molecule</span> <span class="ow">in</span> <span class="n">molecules</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">molecule</span><span class="o">.</span><span class="n">add_fragment</span><span class="p">(</span><span class="n">fragment</span><span class="p">):</span>
                <span class="n">added</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">added</span><span class="p">:</span>
            <span class="n">molecules</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">moleculeClass</span><span class="p">(</span><span class="n">fragment</span><span class="p">,</span> <span class="o">**</span><span class="n">molecule_class_args</span> <span class="p">))</span>

        <span class="n">added_fragments</span><span class="o">+=</span><span class="mi">1</span>

        <span class="k">if</span> <span class="n">added_fragments</span><span class="o">&gt;</span><span class="n">check_eject_every</span><span class="p">:</span>
            <span class="n">current_chrom</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">current_position</span> <span class="o">=</span> <span class="n">fragment</span><span class="o">.</span><span class="n">get_span</span><span class="p">()</span>
            <span class="n">to_pop</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">m</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">molecules</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">can_be_yielded</span><span class="p">(</span><span class="n">current_chrom</span><span class="p">,</span><span class="n">current_position</span><span class="p">):</span>
                    <span class="n">to_pop</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">to_pop</span><span class="p">):</span>
                <span class="k">yield</span> <span class="n">molecules</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="n">j</span><span class="p">)</span>

    <span class="c1"># Yield remains</span>
    <span class="k">yield from</span> <span class="nb">iter</span><span class="p">(</span><span class="n">molecules</span><span class="p">)</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Buys de Barbanson, Jake Yeung.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>