<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>singlecellmultiomics.bamProcessing.bamFeatureDensityVisualisation &#8212; SingleCellMultiOmics 0.0.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for singlecellmultiomics.bamProcessing.bamFeatureDensityVisualisation</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.lines</span> <span class="k">as</span> <span class="nn">mlines</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">singlecellmultiomics.features</span>


<div class="viewcode-block" id="distance_to_hit"><a class="viewcode-back" href="../../../source/singlecellmultiomics.bamProcessing.html#singlecellmultiomics.bamProcessing.bamFeatureDensityVisualisation.distance_to_hit">[docs]</a><span class="k">def</span> <span class="nf">distance_to_hit</span><span class="p">(</span> <span class="n">lookup_coordinate</span><span class="p">,</span> <span class="n">hit_start</span><span class="p">,</span> <span class="n">hit_end</span><span class="p">,</span> <span class="n">hit_strand</span><span class="p">,</span> <span class="n">lookup_strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">distance_from_begin</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">distance</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">distance_from_begin</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">hit_strand</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">:</span>
            <span class="n">distance</span> <span class="o">=</span> <span class="n">lookup_coordinate</span> <span class="o">-</span> <span class="n">hit_start</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">distance</span> <span class="o">=</span>  <span class="n">hit_start</span> <span class="o">-</span> <span class="n">lookup_coordinate</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">hit_strand</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">:</span>
            <span class="n">distance</span> <span class="o">=</span> <span class="n">lookup_coordinate</span> <span class="o">-</span> <span class="n">hit_end</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">distance</span> <span class="o">=</span>  <span class="n">hit_end</span> <span class="o">-</span> <span class="n">lookup_coordinate</span>

    <span class="k">return</span> <span class="n">distance</span></div>

<span class="c1"># check if lookup coordinate falls within a feature:</span>
<div class="viewcode-block" id="distance_to_feature_start"><a class="viewcode-back" href="../../../source/singlecellmultiomics.bamProcessing.html#singlecellmultiomics.bamProcessing.bamFeatureDensityVisualisation.distance_to_feature_start">[docs]</a><span class="k">def</span> <span class="nf">distance_to_feature_start</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span><span class="n">lookup_coordinate</span><span class="p">,</span><span class="n">feature_container</span><span class="p">,</span><span class="n">lookup_strand</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">distance</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">min_distance</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">feature_container</span><span class="o">.</span><span class="n">findFeaturesAt</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="n">chromosome</span><span class="p">,</span><span class="n">lookupCoordinate</span><span class="o">=</span><span class="n">lookup_coordinate</span><span class="p">,</span><span class="n">strand</span><span class="o">=</span><span class="n">lookup_strand</span><span class="p">):</span>
        <span class="n">hit_start</span><span class="p">,</span> <span class="n">hit_end</span><span class="p">,</span> <span class="n">hit_id</span><span class="p">,</span> <span class="n">hit_strand</span><span class="p">,</span> <span class="n">hit_ids</span> <span class="o">=</span> <span class="n">hit</span>
        <span class="c1"># calculate distance relative to selected point</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="n">distance_to_hit</span><span class="p">(</span> <span class="n">lookup_coordinate</span><span class="p">,</span> <span class="n">hit_start</span><span class="p">,</span> <span class="n">hit_end</span><span class="p">,</span> <span class="n">hit_strand</span><span class="p">,</span> <span class="n">lookup_strand</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">min_distance</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">abs</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span><span class="o">&lt;</span><span class="nb">abs</span><span class="p">(</span><span class="n">min_distance</span><span class="p">):</span>
            <span class="n">min_distance</span> <span class="o">=</span> <span class="n">distance</span>
    <span class="k">if</span> <span class="n">distance</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">feature_container</span><span class="o">.</span><span class="n">findNearestFeature</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="n">chromosome</span><span class="p">,</span><span class="n">lookupCoordinate</span><span class="o">=</span><span class="n">lookup_coordinate</span><span class="p">,</span><span class="n">strand</span><span class="o">=</span><span class="n">lookup_strand</span><span class="p">):</span>
            <span class="n">hit_start</span><span class="p">,</span> <span class="n">hit_end</span><span class="p">,</span> <span class="n">hit_id</span><span class="p">,</span> <span class="n">hit_strand</span><span class="p">,</span> <span class="n">hit_ids</span> <span class="o">=</span> <span class="n">hit</span>

            <span class="n">distance</span> <span class="o">=</span> <span class="n">distance_to_hit</span><span class="p">(</span> <span class="n">lookup_coordinate</span><span class="p">,</span> <span class="n">hit_start</span><span class="p">,</span> <span class="n">hit_end</span><span class="p">,</span> <span class="n">hit_strand</span><span class="p">,</span> <span class="n">lookup_strand</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">min_distance</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">abs</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span><span class="o">&lt;</span><span class="nb">abs</span><span class="p">(</span><span class="n">min_distance</span><span class="p">):</span>
                <span class="n">min_distance</span> <span class="o">=</span> <span class="n">distance</span>

    <span class="k">return</span> <span class="n">distance</span></div>

<div class="viewcode-block" id="bam_to_histogram"><a class="viewcode-back" href="../../../source/singlecellmultiomics.bamProcessing.html#singlecellmultiomics.bamProcessing.bamFeatureDensityVisualisation.bam_to_histogram">[docs]</a><span class="k">def</span> <span class="nf">bam_to_histogram</span><span class="p">(</span><span class="n">bam_path</span><span class="p">,</span> <span class="n">add_to</span><span class="p">,</span> <span class="n">feature_container</span><span class="p">,</span> <span class="n">site_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">min_mq</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="mi">15</span><span class="n">_000</span><span class="p">,</span> <span class="n">head</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">quick</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">histogram</span> <span class="o">=</span> <span class="n">add_to</span>
    <span class="k">with</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bam_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">i</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">head</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">i</span><span class="o">&gt;</span><span class="n">head</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">mapping_quality</span><span class="o">&lt;</span><span class="n">min_mq</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="n">library</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;LY&#39;</span><span class="p">)</span>
            <span class="n">chromosome</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_name</span>
            <span class="k">if</span> <span class="n">quick</span> <span class="ow">or</span> <span class="n">site_mode</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">site_mode</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s1">&#39;DS&#39;</span><span class="p">):</span>
                        <span class="k">continue</span>
                    <span class="n">i</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="n">distance</span> <span class="o">=</span> <span class="n">distance_to_feature_start</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span>
                                              <span class="nb">int</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;DS&#39;</span><span class="p">)),</span>
                                              <span class="n">feature_container</span><span class="o">=</span><span class="n">feature_container</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">distance</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>  <span class="nb">abs</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span><span class="o">&lt;</span><span class="n">max_distance</span><span class="p">:</span>
                        <span class="n">histogram</span><span class="p">[</span><span class="n">library</span><span class="p">][</span> <span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">distance</span><span class="o">/</span><span class="n">bin_size</span><span class="p">)</span><span class="o">*</span><span class="n">bin_size</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">i</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="k">for</span> <span class="n">distance</span> <span class="ow">in</span> <span class="p">[</span>
                        <span class="n">distance_to_feature_start</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_start</span><span class="p">,</span> <span class="n">feature_container</span><span class="o">=</span><span class="n">feature_container</span><span class="p">),</span>
                        <span class="n">distance_to_feature_start</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">feature_container</span><span class="o">=</span><span class="n">feature_container</span><span class="p">)</span>
                    <span class="p">]:</span>

                        <span class="k">if</span> <span class="n">distance</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>  <span class="nb">abs</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span><span class="o">&lt;</span><span class="n">max_distance</span><span class="p">:</span>
                            <span class="n">histogram</span><span class="p">[</span><span class="n">library</span><span class="p">][</span> <span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">distance</span><span class="o">/</span><span class="n">bin_size</span><span class="p">)</span><span class="o">*</span><span class="n">bin_size</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">q_pos</span><span class="p">,</span> <span class="n">ref_pos</span> <span class="ow">in</span> <span class="n">read</span><span class="o">.</span><span class="n">get_aligned_pairs</span><span class="p">(</span><span class="n">matches_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">with_seq</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
                    <span class="n">distance</span> <span class="o">=</span> <span class="n">distance_to_feature_start</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">ref_pos</span><span class="p">,</span> <span class="n">feature_container</span><span class="o">=</span><span class="n">feature_container</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">distance</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">abs</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span><span class="o">&lt;</span><span class="n">max_distance</span><span class="p">:</span>
                        <span class="n">histogram</span><span class="p">[</span><span class="n">library</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">distance</span><span class="o">/</span><span class="n">bin_size</span><span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{bam_path}</span><span class="s1"> read </span><span class="si">{i}</span><span class="s1"> reads&#39;</span><span class="p">)</span></div>

<span class="k">if</span> <span class="vm">__name__</span><span class="o">==</span><span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">160</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">&#39;Agg&#39;</span><span class="p">)</span>
    
    <span class="n">argparser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
     <span class="n">formatter_class</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span><span class="p">,</span>
     <span class="n">description</span><span class="o">=</span><span class="s1">&#39;Visualise feature density of a bam file. (Coverage around stop codons, start codons, genes etc)&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;output plot folder path, every library will be visualised separately&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;./plots/&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-d&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;output data folder path, the data used for plotting will be exported to this folder&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;./tables/&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-features&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;stop_codon,start_codon&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;features to plot, separate by comma without space&quot;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;alignmentfiles&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s1">&#39;*&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-gtf&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;GTF file containing the features to plot&quot;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-head&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">100</span><span class="n">_000_000</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Use this amount of reads per bam file (or less if the bam file has less reads)&quot;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--bySite&#39;</span><span class="p">,</span>  <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Use the DS tag to count density&quot;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--binSize&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span><span class="n">default</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--maxDistance&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span><span class="n">default</span><span class="o">=</span><span class="mi">15</span><span class="n">_000</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Size of the window&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--minMQ&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span><span class="n">default</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Minimum mapping quality&#39;</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">argparser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">features</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">o</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">o</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">d</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">d</span><span class="p">)</span>

    <span class="c1"># Load the GTF files with the desired annotations</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
        <span class="n">annotations</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span> <span class="o">=</span> <span class="n">singlecellmultiomics</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">FeatureContainer</span><span class="p">()</span>
        <span class="n">annotations</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span><span class="o">.</span><span class="n">loadGTF</span><span class="p">(</span> <span class="n">args</span><span class="o">.</span><span class="n">gtf</span><span class="p">,</span> <span class="n">select_feature_type</span><span class="o">=</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span> <span class="p">)</span>


    <span class="n">histograms</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span>
                    <span class="k">lambda</span><span class="p">:</span><span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">))</span> <span class="c1"># feature-&gt;library-&gt;hist</span>
    <span class="c1"># Create histogram per feature / library</span>
    <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">bam_path</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">alignmentfiles</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Now reading </span><span class="si">{bam_path}</span><span class="s2"> for annotation type </span><span class="si">{feature}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">bam_to_histogram</span><span class="p">(</span>
                <span class="n">bam_path</span><span class="p">,</span>
                <span class="n">histograms</span><span class="p">[</span><span class="n">feature</span><span class="p">],</span>
                <span class="n">annotations</span><span class="p">[</span><span class="n">feature</span><span class="p">],</span>
                <span class="n">site_mode</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">bySite</span><span class="p">,</span>
                <span class="n">bin_size</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">binSize</span><span class="p">,</span>
                <span class="n">max_distance</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">maxDistance</span><span class="p">,</span>
                <span class="n">min_mq</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">minMQ</span><span class="p">,</span>
                <span class="n">head</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">head</span><span class="p">)</span>

    <span class="c1"># Write</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Writing tables&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">feature</span><span class="p">,</span> <span class="n">feature_data</span> <span class="ow">in</span> <span class="n">histograms</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">library</span><span class="p">,</span> <span class="n">histogram</span> <span class="ow">in</span> <span class="n">feature_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span> <span class="nb">list</span><span class="p">(</span><span class="n">histogram</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span><span class="nb">list</span><span class="p">(</span><span class="n">histogram</span><span class="o">.</span><span class="n">values</span><span class="p">())])</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
            <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;bin&#39;</span><span class="p">,</span><span class="s1">&#39;observations&#39;</span><span class="p">]</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{args.d}</span><span class="s1">/</span><span class="si">{feature}</span><span class="s1">_</span><span class="si">{library}</span><span class="s1">_</span><span class="si">{args.maxDistance}</span><span class="s1">_</span><span class="si">{args.binSize}</span><span class="s1">.csv&#39;</span><span class="p">)</span>

    <span class="c1"># Plot the histograms</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Plotting&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">feature</span><span class="p">,</span> <span class="n">feature_data</span> <span class="ow">in</span> <span class="n">histograms</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">library</span><span class="p">,</span> <span class="n">histogram</span> <span class="ow">in</span> <span class="n">feature_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{feature}</span><span class="s1"> </span><span class="si">{library}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span> <span class="nb">list</span><span class="p">(</span><span class="n">histogram</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span><span class="nb">list</span><span class="p">(</span><span class="n">histogram</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span> <span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Distance from </span><span class="si">{feature}</span><span class="s1"> codon [bp]&#39;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;# bases&#39;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{library}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=-</span><span class="mi">10</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{args.o}</span><span class="s1">/</span><span class="si">{feature}</span><span class="s1">_</span><span class="si">{library}</span><span class="s1">_</span><span class="si">{args.maxDistance}</span><span class="s1">_</span><span class="si">{args.binSize}</span><span class="s1">.png&#39;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Buys de Barbanson, Jake Yeung.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>