<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>singlecellmultiomics.bamProcessing.bamPlotRTstats &#8212; SingleCellMultiOmics 0.0.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for singlecellmultiomics.bamProcessing.bamPlotRTstats</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">160</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">&#39;Agg&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.lines</span> <span class="k">as</span> <span class="nn">mlines</span>

<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">pysamiterators.iterators</span> <span class="k">as</span> <span class="nn">pyts</span>
<span class="kn">import</span> <span class="nn">singlecellmultiomics.universalBamTagger.universalBamTagger</span> <span class="k">as</span> <span class="nn">ut</span>
<span class="kn">import</span> <span class="nn">importlib</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">itertools</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">singlecellmultiomics.features</span>


<div class="viewcode-block" id="nlaIII_molecule_acceptance_function"><a class="viewcode-back" href="../../../source/singlecellmultiomics.bamProcessing.html#singlecellmultiomics.bamProcessing.bamPlotRTstats.nlaIII_molecule_acceptance_function">[docs]</a><span class="k">def</span> <span class="nf">nlaIII_molecule_acceptance_function</span><span class="p">(</span><span class="n">molecule</span><span class="p">):</span>
    <span class="n">first_read</span> <span class="o">=</span> <span class="n">molecule</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">first_read</span><span class="o">.</span><span class="n">mapping_quality</span><span class="o">&lt;</span><span class="mi">60</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="n">reject</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">first_read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s1">&#39;XA&#39;</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">alt_align</span> <span class="ow">in</span> <span class="n">first_read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;XA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt_align</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span> <span class="c1"># Sometimes this tag is empty for some reason</span>
                <span class="k">continue</span>

            <span class="n">hchrom</span><span class="p">,</span> <span class="n">hpos</span><span class="p">,</span> <span class="n">hcigar</span><span class="p">,</span> <span class="n">hflag</span> <span class="o">=</span> <span class="n">alt_align</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">hchrom</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;_alt&#39;</span><span class="p">):</span>
                <span class="n">reject</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
    <span class="k">return</span> <span class="n">reject</span></div>
<span class="k">if</span> <span class="vm">__name__</span><span class="o">==</span><span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">argparser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
         <span class="n">formatter_class</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span><span class="p">,</span>
         <span class="n">description</span><span class="o">=</span><span class="s1">&#39;Visualise feature density of a bam file. (Coverage around stop codons, start codons, genes etc)&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;bamFile&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-head&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Process this many molecules&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-binSize&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-maxfs&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">900</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;X axis limit of fragment size plot&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-maxOverseq&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--notstrict&#39;</span><span class="p">,</span>  <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">)</span>
    <span class="n">argparser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span>  <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;RT_dist&#39;</span><span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">argparser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>


    <span class="k">def</span> <span class="nf">dd</span><span class="p">():</span>
        <span class="k">return</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span> <span class="p">)</span>

    <span class="n">fragment_distribution</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">()</span>
    <span class="n">fragment_distribution_raw</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span> <span class="p">)</span> <span class="c1"># overseq -&gt; fragmentisze -&gt; counts</span>
    <span class="c1"># Read size fragments</span>
    <span class="n">fragment_distribution_raw_rf</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span> <span class="n">dd</span> <span class="p">)</span><span class="c1"># lib -&gt; overseq -&gt; fragmentisze (span) -&gt; counts</span>
    <span class="n">gc_distribution</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">()</span>
    <span class="n">gc_frag_distribution</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span> <span class="p">)</span> <span class="c1"># fragment size -&gt; observed gc/at+gc ratio</span>
    <span class="c1"># fragmentSize -&gt; umi obs</span>

    <span class="n">rt_frag_distribution</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span> <span class="p">)</span>
    <span class="c1"># fragment size -&gt; amount of RT reactions -&gt; count</span>

    <span class="n">observed_cuts</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">()</span>
    <span class="n">used</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">used_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">gc_capture</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">with</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bamFile</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">molecule</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span> <span class="n">ut</span><span class="o">.</span><span class="n">MoleculeIterator</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">umi_hamming_distance</span><span class="o">=</span><span class="mi">1</span>
                                                        <span class="p">)):</span>



            <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">notstrict</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">nlaIII_molecule_acceptance_function</span><span class="p">(</span><span class="n">molecule</span><span class="p">):</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">head</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">used</span><span class="o">&gt;</span><span class="n">args</span><span class="o">.</span><span class="n">head</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Stoppping, saw enough molecules&#39;</span><span class="p">)</span>
                <span class="k">break</span>

            <span class="n">rt_reactions</span> <span class="o">=</span> <span class="n">ut</span><span class="o">.</span><span class="n">molecule_to_random_primer_dict</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>
            <span class="n">amount_of_rt_reactions</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rt_reactions</span><span class="p">)</span>


            <span class="c1">#this obtains the maximum fragment size:</span>
            <span class="n">frag_chrom</span><span class="p">,</span> <span class="n">frag_start</span><span class="p">,</span> <span class="n">frag_end</span> <span class="o">=</span> <span class="n">pyts</span><span class="o">.</span><span class="n">getListSpanningCoordinates</span><span class="p">([</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">])</span>

            <span class="c1">#Obtain the fragment sizes of all RT reactions:</span>
            <span class="n">rt_sizes</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">rt_end</span><span class="p">,</span><span class="n">hexamer</span><span class="p">),</span> <span class="n">fragment</span> <span class="ow">in</span> <span class="n">rt_reactions</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">rt_chrom</span><span class="p">,</span> <span class="n">rt_start</span><span class="p">,</span> <span class="n">rt_end</span> <span class="o">=</span> <span class="n">pyts</span><span class="o">.</span><span class="n">getListSpanningCoordinates</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">fragment</span><span class="p">))</span>
                <span class="n">rt_sizes</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">rt_end</span><span class="o">-</span><span class="n">rt_start</span><span class="p">])</span>


            <span class="n">first_read</span> <span class="o">=</span> <span class="n">molecule</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">site</span> <span class="o">=</span> <span class="n">first_read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;DS&#39;</span><span class="p">)</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="n">first_read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;RS&#39;</span><span class="p">)</span>
            <span class="n">library</span> <span class="o">=</span> <span class="n">first_read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;LY&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gc_capture</span><span class="p">:</span>
                <span class="n">sequence</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">frag_chrom</span><span class="p">,</span> <span class="n">frag_start</span><span class="p">,</span> <span class="n">frag_end</span><span class="p">)</span>
                <span class="n">gc</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">)</span><span class="o">+</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;G&#39;</span><span class="p">)</span>
                <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
            <span class="n">used</span> <span class="o">+=</span><span class="mi">1</span>

            <span class="k">if</span> <span class="kc">False</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">:</span>

                    <span class="n">c</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span> <span class="n">ref_sites</span><span class="p">,</span> <span class="n">site</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span> <span class="c1">#+ means goes left</span>
                    <span class="n">cut_start</span> <span class="o">=</span>  <span class="n">ref_sites</span><span class="p">[</span><span class="n">c</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                    <span class="n">cut_end</span>  <span class="o">=</span>  <span class="n">site</span>

                <span class="k">else</span><span class="p">:</span>

                    <span class="n">c</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span> <span class="n">ref_sites</span><span class="p">,</span> <span class="n">site</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
                    <span class="n">cut_start</span> <span class="o">=</span>  <span class="n">site</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">cut_end</span>  <span class="o">=</span>  <span class="n">ref_sites</span><span class="p">[</span><span class="n">c</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="n">cut_end</span> <span class="o">=</span> <span class="n">ref_sites</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="c1"># this happens at the last site at the chromosome</span>

                <span class="n">fragment_size</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">cut_start</span><span class="o">-</span><span class="n">cut_end</span><span class="p">)</span>
                <span class="n">fragment_distribution</span><span class="p">[</span><span class="n">fragment_size</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
            <span class="k">if</span> <span class="n">gc_capture</span><span class="p">:</span>
                <span class="n">gc_distribution</span><span class="p">[</span><span class="n">gc</span><span class="o">/</span><span class="n">length</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">gc_frag_distribution</span><span class="p">[</span><span class="n">fragment_size</span><span class="p">][</span><span class="n">gc</span><span class="o">/</span><span class="n">length</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">fragment_distribution_raw</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">molecule</span><span class="p">)][</span><span class="n">fragment_size</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">rt_sizes</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
                <span class="n">mean_rt_size</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">mean_rt_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">rt_sizes</span><span class="p">))</span>
            <span class="n">fragment_distribution_raw_rf</span><span class="p">[</span><span class="n">library</span><span class="p">][</span><span class="nb">len</span><span class="p">(</span><span class="n">molecule</span><span class="p">)][</span><span class="n">mean_rt_size</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">rt_frag_distribution</span><span class="p">[</span><span class="n">mean_rt_size</span><span class="p">][</span><span class="nb">len</span><span class="p">(</span><span class="n">rt_reactions</span><span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">used_reads</span><span class="o">+=</span><span class="nb">len</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>


    <span class="n">bin_size</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">binSize</span>
    <span class="n">m_overseq</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">maxOverseq</span>

    <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{args.o}</span><span class="s1">_raw_data.pickle.gz&#39;</span><span class="p">,</span><span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fo</span><span class="p">:</span>

        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">fragment_distribution_raw_rf</span><span class="p">,</span> <span class="n">fo</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">library</span> <span class="ow">in</span> <span class="n">fragment_distribution_raw_rf</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span><span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">7</span><span class="p">))</span>
        <span class="n">ax</span>  <span class="o">=</span> <span class="n">axes</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Read fragment size distribution</span><span class="se">\n</span><span class="si">{used}</span><span class="s1"> molecules / </span><span class="si">{used_reads}</span><span class="s1"> fragments  analysed</span><span class="se">\n</span><span class="si">{library}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">table</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">overseq</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">m_overseq</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1">#Rebin in 10bp bins:</span>
                <span class="n">rebinned</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">f_size</span><span class="p">,</span> <span class="n">obs</span> <span class="ow">in</span> <span class="n">fragment_distribution_raw_rf</span><span class="p">[</span><span class="n">library</span><span class="p">][</span><span class="n">overseq</span><span class="p">]</span><span class="o">.</span><span class="n">most_common</span><span class="p">():</span>
                    <span class="n">rebinned</span><span class="p">[</span> <span class="nb">int</span><span class="p">(</span> <span class="n">f_size</span><span class="o">/</span><span class="n">bin_size</span><span class="p">)</span><span class="o">*</span><span class="n">bin_size</span> <span class="p">]</span> <span class="o">+=</span> <span class="n">obs</span>
                <span class="n">obs_dist_fsize</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">rebinned</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
                <span class="n">obs_dist_freq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">rebinned</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
                <span class="n">sorting_order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_dist_fsize</span><span class="p">)</span>
                <span class="n">obs_dist_fsize</span> <span class="o">=</span> <span class="n">obs_dist_fsize</span><span class="p">[</span><span class="n">sorting_order</span><span class="p">]</span>
                <span class="n">obs_dist_freq</span> <span class="o">=</span> <span class="n">obs_dist_freq</span><span class="p">[</span><span class="n">sorting_order</span><span class="p">]</span>
                <span class="n">obs_dist_density</span> <span class="o">=</span> <span class="n">obs_dist_freq</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">obs_dist_freq</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">obs_dist_fsize</span><span class="p">):</span>
                    <span class="n">table</span><span class="p">[(</span><span class="n">overseq</span><span class="p">,</span><span class="n">x</span><span class="p">)]</span> <span class="o">=</span><span class="p">{</span><span class="s1">&#39;obs_dist_freq&#39;</span><span class="p">:</span><span class="n">obs_dist_freq</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                        <span class="s1">&#39;obs_dist_density&#39;</span><span class="p">:</span><span class="n">obs_dist_density</span><span class="p">[</span><span class="n">i</span><span class="p">]}</span>

                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">obs_dist_fsize</span><span class="p">,</span> <span class="n">obs_dist_density</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="p">(</span><span class="n">overseq</span><span class="o">/</span><span class="n">m_overseq</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span><span class="n">label</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{overseq}</span><span class="s1"> read fragments / umi&#39;</span> <span class="p">)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span><span class="n">args</span><span class="o">.</span><span class="n">maxfs</span><span class="p">)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;fragment size&#39;</span><span class="p">)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;density&#39;</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
                <span class="k">pass</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{args.o}</span><span class="s1">_</span><span class="si">{library}</span><span class="s1">.png&#39;</span><span class="p">)</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">table</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{args.o}</span><span class="s1">_</span><span class="si">{library}</span><span class="s1">.csv&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">pass</span>

    <span class="c1"># Export the table:</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Buys de Barbanson, Jake Yeung.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>